readme
================================================
+                   README                     +
================================================

StrangeBrew Java - 2.0.2
(c)2000-2008 Drew Avis - drew.avis@gmail.com

Thank you for downloading and trying StrangeBrew Java.

This file contains:
* What is this?
* Recipes at bodensatz.com
* Running StrangeBrew Java
* Extra help for running on Linux
* Setting the look and feel
* Files in this release
* Required JRE
* Known Issues (bugs)
* Reporting Bugs
* What's New in this release?
* Roadmap
* Building the project
* Contributors
* Copyrights and Licenses

===============
 What is this?
===============

This is a "release quality" release of StrangeBrew Java, which means 
it's feature-complete, and is pretty stable.  If you find bugs, please
report them via the Sourceforge bug tracker, or email 
bugs@strangebrew.ca.

StrangeBrew Java is a Java port of the popular Windows homebrewing
software, StrangeBrew (www.strangebrew.ca).

Donations are accepted on the SourceForge project homepage.

NOTE: We've changed the numbering of Java releases to 2.0, based on 
the idea that this is a continuation of the Windows version 1.8.


========================
Recipes at bodensatz.com
========================

There is a repository of StrangeBrew recipes available for viewing and
download at www.bodensatz.com/sb, hosted by Alan McKay.  Please feel free
to download recipes to open with StrangeBrew Java.  

==========================
 Running StrangeBrew Java
==========================

Running StrangeBrew:
1. Unzip the sb_XXX.zip file (where XXX is the latest version, eg: "sb_2.0.1").
Be sure to preserve the directory structure.
2. Type "java -jar strangebrew.jar" at the command line.  Under Windows,
you can right-click this file, select "Open With > " and select a Java
installation. You can also just double-click the file, if ".jar" files
are associated with javaw.exe.  Double-clicking reportedly also works on
Mac OSX.  There's probably a Linux way to do this, but I don't know what it is.

NOTE: StrangeBrew loads data (ingredients, styles, etc) from:
src/ca/strangebrew/data/
When you unzip the file, be sure to preserve directory names, so
that StrangeBrew can find this data.

===============================
Extra help for running on Linux
===============================

These are notes from Scott Alfter via rec.crafts.brewing:

1) Make sure CONFIG_BINFMT_MISC is enabled in the kernel (built-in or
   module)...look in Executable file formats/Emulations -> Kernel support
   for MISC binaries.
2) Add the following to /etc/fstab:
   binfmt_misc /proc/sys/fs/binfmt_misc binfmt_misc defaults 0 0
3) Create /usr/bin/jexec with the following contents and make it executable
   by everybody:
   #!/bin/sh
   java -jar $*
4) Add the following to whatever script your distro provides for local
   customization (in Gentoo, it's /etc/conf.d/local.start):
   echo ':jarexec:M::PK\x03\x04::/usr/bin/jexec:' >/proc/sys/fs/binfmt_misc/register

When you reboot (or if you execute the preceding from a root prompt), you
can chmod 755 strangeBrew_fat.jar and then run it with the following:

./strangeBrew_fat.jar

NOTE: At least one person has experienced problems running StrangeBrew on Fedora
using their 1.4.2 JRE, which is the GNU gcj.  The problem was solved by using the
1.5 JRE from Sun.  

For Ubuntu:
Install Sun's JRE:
	$ sudo apt-get install sun-java6-jre
Update default system JVM:
	$ sudo update-alternatives --config java
  Select java-6 from the list
	$ sudo update-alternatives --config jar
  Select java-6 from the list

See https://help.ubuntu.com/community/Java for more help with Ubuntu.

Greg LaPolla points out:
When you create a launcher on the desktop and start the app with the java -jar 
command it displays the "can't find the database" error.  You can create a shell
script to fix it by simply changing to the StrangeBrew directory first:

#!/bin/sh
cd ~/StrangeBrew
java -jar StrangeBrew.jar

=========================
Setting the Look and Feel
=========================

The Java Swing library (which StrangeBrew uses for its user interface) supports
a "pluggable look and feel" (LAF) -- meaning you can easily change the appearance 
of Swing applications to suit your own tastes.  

(Some betas of StrangeBrew used the JGoodies Looks LAF, but I removed
it as the default to give users more control over how the application looks).

Very nice look and feels are available on-line from several sources.  Here are 
a few of my favourites for StrangeBrew:
* JGoodies Looks: http://javootoo.l2fprod.com/plaf/jgoodies/index.php
* Liquid L&F: https://liquidlnf.dev.java.net/
* Synthetica L&F: http://www.javasoft.de/jsf/public/products/synthetica
* Tonic L&F: http://www.digitprop.com/p.php?page=toniclf

Download the .jar file you'd like to try, and put it in your Java home directory under
/lib/ext.

You can set the Look and Feel in a couple of ways:

1. From the Command-line:
Use the -Dswing.defaultlaf= option when starting the application.
Here's an example of starting StrangeBrew using the Synthetica LAF:
java -Dswing.defaultlaf=de.javasoft.plaf.synthetica.SyntheticaStandardLookAndFeel -jar strangebrew.jar

The "de.javasoft.plaf.synthetica.SyntheticaStandardLookAndFeel" bit is the full class 
name of the Look And Feel -- you'll have to read the documentation for each LAF to find 
out the proper text to put here

2. Using swing.properties:
You can set the default LAF for all Java Swing applications by editing the swing.properties
file in your $JAVA_HOME/lib directory.  You need to add two lines for your new LAF, and
change the default line.  Here's an example of the new & changed lines using the Tonic
LAF:
swing.installedlaf.tonic.name = Tonic
swing.installedlaf.tonic.class = com.digitprop.tonic.TonicLookAndFeel
swing.defaultlaf=com.digitprop.tonic.TonicLookAndFeel

Now start StrangeBrew as you would normally, and it should use the new LAF.

==========
Files 
==========

The file you downloaded should contain these files (at least):
  * strangebrew.jar: a Fat Jar file which you can run (see above).  This
    file also contains the complete Java source for the application.  To
    view the source, decompress this file with an archive utility.
  * gpl.txt: the GPL license this software is distributed under
  * readme.txt: this file
  * src/ca/strangebrew/data: data files and the xslt files required 
    for exporting HTML and printing.  You can modify the xslt files if 
    you want to customize the look of the HTML.
  * help/: help files.  These files are under development.
  * recipes/: example recipe files in various formats.    
  * src/*: all the Java files you need to build and run StrangeBrew yourself.  
    You will need to either download or extract a couple of other .jar files 
    that StrangeBrew requires.


============
Required JRE
============

You require a Java Runtime Environment (JRE) on your system to run
StrangeBrew.  It requires versions 1.5 and up (See 
http://java.sun.com/j2se/1.5.0/index.jsp). As of version 2.0.1, 
StrangeBrew no longer runs under version 1.4.   StrangeBrew has
been tested and runs on Linux, MacOS-X, and Windows.  

============
Known Issues
============

* When importing recipes from StrangeBrew for Windows (1.8 or earlier),
fermentables that are sugars are incorrectly considered "grains".  Their
gravity contribution is calculated using the recipe efficiency, rather than
100%.  You should uncheck the "M" (for mashed) column, which indicates that
the ingredient is a sugar.  This applies to all sugar-type fermentables, such
as honey, DME, LME, maple syrup, and sugars.

* There is no way to add or modify ingredients from within SB Java. Currently 
any custom data for hops, malts, water profiles, yeast, salts and primeing sugars
must be entered by hand into the cvs files. *WARNING* These cvs files will be 
over writen with each new install.

==============
Reporting Bugs
==============

We're using the Sourceforge tracking feature to track bugs.  Please go to 
www.sourceforge.net/projects/strangebrew and click on "bugs" to view a list
of reported bugs.  If you've found one that's not there, please report it!

==========
What's New
==========

2.0.2
--------
New Features:
- You can now specify the BJCP style year you want to use (2004 or 2008).
- Table column sizes are now remembered when you close and restart the application
  (Request 1493295).
- Diluted recipe feature works a lot better.  Diluted specs show up on the main recipe
  details tab.  
- If StrangeBrew can't find the ingredient databases, it will ask you where they are instead
  of silently starting anyway.  
- StrangeBrew now looks more like a native MacOS/X application on Macs.  
- Windows Installer

Bugs Fixed:
- Misc water loss is now set correctly on the Water tab.
- Rager formula bug fixed.
- New Recipe Defaults: Temp Units now displays properly.

2.0.1
--------
New Features:
- New Gravity Teperature Tool Dialog (Request 1222300)
- New selectable user locale, which affects date format (Bug 1480514)
- New Fermentation Schedule tab
- New Carbonation tab
- New Water Treatment tab

Bugs Fixed:
- Default hop type (Leaf, Pellet, Plug) is now configurable (Request 1489276) 
- Text field auto selection (Bug 1489252)
- You can now select-by-typing in malt and hop lists (Bug 1493279)
- Mash Steps now insert at end of list (Request 1497970)
- Changed default bottle unit and size to imperal (Bug 1493267) 
- UI improvments to the saved "default" mash schedules


2.0.0
--------
Bugs Fixed:
- Mash ratio and ratio units are now properly displayed.
- After storing ingredients from an imported recipe, the database is no longer corrupted.
- Malt and hop drop-downs behave better (Bugs 1580524 and 1511954).
- Selecting a date now permanently changes it (Bug 1581203).
- The comments scrollbar now behaves correctly.
- File extensions don't keep multiplying every time you open the Save/SaveAs/Export dialogs.
- The Min column in the hops table is now handled correctly.
- The style dropdown now displays from the top of the list, rather than the bottom, for an empty style (Bug 1493269).


2.0.rc1
--------
New Features:
- StrangeBrew no longer uses the JGoodies Look And Feel by default, as this was
  causing some refresh issues on older versions of Mac OSX.  Instead, you can set
  the LAF yourself. (See the "Setting the Look And Feel" section above).
- You can now set the bottom of temp ranges used to determine mashout and sparge
  steps.  The defaults are 161F for mashout, and 170F for sparge.  This means that
  when you add a mashout step, it uses the middle of the range (161 + 170 / 2, or
  165.5F) when you add a mashout step, and if you change the start temp to <161, it
  becomes a different step type.  You can adjust these numbers to suit your mashing
  preferences.
- More and better StrangeBrew Java-specific help.  
- New Print toolbar button.

Bugs Fixed:
- FWH, mash hop, and dry hop IBU values are now correctly calculated.
- The strangebrew.ini file is now saved and read from the correct location on 
  non-Windows machines.
- HTML export now works again on non-Windows machines.  

2.0.b7
--------
New Features:
- You can now apply default mash schedules on the Mash tab to quickly load
  a popular schedule.  You can also save the current schedule to the list of
  defaults.  You can edit the /data/mashdefaults.csv file by hand to add or 
  remove a default schedule.
- The Resize Recipe dialog now allows you to also convert hops and fermentables
  (Bug 1489903).
- New potential extract calculator.  
- You can now add sparge steps to a mash schedule.  Add a step, and change the 
  type to "sparge".  If you add more than one sparge step, it's automatically
  considered a batch sparge, and correct additions are calculated.  
- You can now set the parameters for a thick and thin decoction on the 
  settings tab.
- You can now add cereal mash steps to a mash schedule.  Add a step, and change the
  method to "cereal mash".  Enter the weight of the grain used in the cereal mash.
  Other values will update accordingly.  Note: the temperature of the step is adjusted
  if the cereal mash addition brings the step above the targe temp.  Try
  decreasing the weight and adjusting the temperature until you get the values you
  want.  
- New Refractometer Utility lets you peform various calculations with readings
  from a refractometer.  

Bugs Fixed:
- You can now specify a default boil time on the Preferences Dialog (Bug 1493270)
- Style tab now displays correct SRM value when you're using EBC colour (Bug 1497989)
- You can now set various mash defaults on the Preferences dialog (Bug 1489256)
- You can now set boil units (C or F) as well as boil temp on the Preferences
  dialog (Bug 1489250)
- Cost tab now re-calculates correctly when values change (Bug 1493994)
- Final beer volume shown on cost tab now matches volume shown on water tab (Bug 1493294)
- BU:GU ratio is now displayed correctly on the Style panel (Bug 1493283).
- Dilution panel is now updated when you first open a recipe (Bug 1492475).
- Setting pre-boil amount now correctly updates post-boil amount.

ChangeLog.txt
Change Log:
Jan 14, 2008 - Greg LaPolla
- Removed NSIS Windows Based Installer 
- Added IzPack Java Based Installer now installs natively on Windows and Linux
 
July 7, 2008 - Drew Avis
- Added option to select BJCP style definintion year

June 11, 2008 - Greg LaPolla
- Added New styleguide for 2008 BJCP guidelines
- Added a new build option in build.xml to generate a source distribution zip file

May 28, 2008 - Drew Avis
- Fixed a bug where the misc loss wasn't set correctly.

May 27, 2008 - Greg LaPolla
- Built windows installer using NSIS
- Added NSIS installer scripts and build process to the ant script
- Changed some build dependencies in the ant build script
- Fixed Rager math calculations now using Math.tanh from Java math library
- Fixed Temp Units display issue in Recipe Preferences

Apr 9, 2008 - Drew Avis
- Fixed display of * and version in title, recipe version.
- Added style panel update back to displayRecipe(), to fix update problem.

Mar 10, 2008 - Drew Avis
- Fixing some lost action listener registrations.
- Fixed problem w/ FinalWortVol vs PostBoilVol confusion.

Feb 21, 2008 - Drew Avis
- Added check for MacOS/X to splash, which makes the app look more native on Macs.  (Suggested by
Greg LaPolla).

Jan 26, 2008 - Jim Clark
- Carried re-cacle bug fix into WaterPanel. DilutionPanel may still need work, but it uses JSpinners and needs more investigation first.
- Moved StrangeSwing.round() into SBStringUtils since its sort of the inverse functionality of SBStringUtils.format().
   Ei: you format a double to a string for display and you round a display string to a double.

Jan 22, 2008 - Drew Avis
- Worked on re-calc bug, and fixed a bug where recipes saved using vol units != US gals were importing size
incorrectly (the problem is in Recipe.setVolUnits(), which was converting rather than just setting units). I
think we still need a Recipe.convertVolUnits(), which is called when you change units...

Jan 17-18, 2008 - Jim Clark
- Modified how Recipe stores wort volumes to cut down on the number of variables neededing to be in sync. This
   should provide a cleaner/easier interface to Recipe for DilutedRecipe.
- Reworked StrangeSwing to consolidate all of the event listners and the "displayRecipe()" call to prevent infinite
   recurtion loops as one display calls display recipe which sets a data element with a listner that calls displayRecipe.
   As a side effect, this elimintate a significant amount of extranious event calls coming out of that function. What used
   to happen would end up recurcivly calling event setters all over thoughout the app, now it only calls the ones on the page
   in view. The reset calls are all triggered off using the tab pane now.
- Finished off DilutedRecipe. Works seamlessly thoughout app and all the ugglyness is contained in DilutedRecipe and DilutedPanel.

Jan 16, 2008 - Jim Clark
- Changed many "ArrayList"s to "List"'s to follow best pratices when using collections. Ei: Program in terms of an 
   interface type (List) rather than implementation type (ArrayList). Eg:
 		List list = new ArrayList();
   is better than:
        ArrayList list = new ArrayList();
   This is so later when we want that Array to be a LinkedList or some custom list implementation of our own, everthgin just
   works.
- Started ripping Diluted Recipe stuff out of Recipe. Currently I would say Diluted Recipe is broken beyond imagineing, but I wanted
   to check this in now, since most of the changes left are going to be in DilutedRecipe.java and DilutionPanel.java and shouldn't
   make it a merging nightmare in a week when I am finaly done.

Jan 16, 2008 - Drew Avis
- Added support for saved column widths between application sessions.

Jan 15, 2008 - Drew Avis
- Now we handle the situation where the ingredient DB is moved.
- Adding Ant build file.

Jan 14, 2008 - Jim Clark
- Fixed some bugs with Salts and moved data out of CVS into static array. The data contents are tightly coupled to the salt calculations
   so it didn't make sense to have the salts configurable on the fly when adding a salt still involved some code work to impliement
   this helped clean up the WaterTreatmentPanel too, so overall.. IMHO, better.
- Added gram->tsp conversion into salt addisions.
- Added acid required for 5.2pH target as well as to hit target water pH. Some people like to emulate water, some like 5.2.

Jan 9, 2008 - Jim Clark
- Added conversion dialog to menu (wopse - missed 2.0.1)

Jan 7, 2008 - Jim Clark
- Fixed bug with the "isDirty" flag and the Carb Panel so that all recipes arn't always dirty or in need of being saved all the time
- Fixed Salt toString functions.

Jan 7, 2008 - Drew Avis
- Fixed broken format string in WaterProfile.toString().
- Fixed a bug on the Note tab where notes were being lost.
- Made debug output optional (w/ -d command line option)

Jan 5, 2008 - Jim Clark
- Down to 1 Type safty warrnings by implemeing various Collection<Type> stuff in various places.
   The one remaining is due to calling Generics unfrinedly SWING code.
- Moved prime sugar and keg presure calcs into Recipe from UI
- Moved water treatment/acid/salt calcs into recipe
- Added acid to save/load/xlst printing
- Acid pulls values from source and target water (perhaps a bad thing?)
- Changed some class comparisons to use instanceof keyword instead of string compareing the class name

Jan 3-4, 2008 - Drew Avis
- Added some tooltips for toolbar buttons
- Added copy to clipboard toolbar button
- Fixed bug where a new hop had Add and Type switched
- You can now set target vol on the carbonation tab
- Tried to get priming sugar amount to re-calculate when you set targetVol on the recipe, but gave up
- Cleaned up some Java 1.6 things (String.isEmpty)
- Made a few changes to reduce compiler warnings (use of raw ArrayList)

Dec 28, 2007 - Jim Clark
- Consolidated a lot of fixed strings scatetred though out app into static strings to

src/strangebrew/Mash.java
package strangebrew;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.Collections;

/**
 * $Id: Mash.java,v 1.12 2005/06/10 20:47:03 andrew_avis Exp $
 * @author aavis
 *
 */

public class Mash {
	
	// set this:
	private double maltWeightLbs;
	
	//options:
	private double mashRatio;
	private String mashRatioU;
	private String tempUnits;
	private String volUnits;
	private double grainTemp;
	private String mashTempU;
	
	// calculated:
	private double volQts;
	private int totalTime;
	private double absorbedQTS;
	private double spargeTotalQTS;
	private double totalWaterQTS;
	private double chillShrink;
	
	
	// steps:
	private ArrayList steps = new ArrayList();
	
	// format:
	public DecimalFormat df1 = new DecimalFormat("#####0.0");
	
	public Mash(){

		Options opts = new Options("mash");
		 mashRatio = opts.getDProperty("optMashRatio");
		 mashRatioU = opts.getProperty("optMashRatioU");;
		 tempUnits = opts.getProperty("optMashTempU");
		 volUnits = opts.getProperty("optMashVolU");
		 grainTemp = opts.getDProperty("optGrainTemp");
	}

	public class MashStep {
		private String type;
		private double startTemp;
		private double endTemp;
		private String tempU;
		private String method;
		private int minutes;
		private int rampMin;
		private String directions;

		public Quantity infuseVol = new Quantity();
		public Quantity decoctVol = new Quantity();

		public MashStep(String t, double st, double et, String tu, String m, int min,
				int rmin) {
			type = t;
			startTemp = st;
			endTemp = et;
			tempU = tu;
			method = m;
			minutes = min;
			rampMin = rmin;
		}

		// default constructor:
		public MashStep() {
			rampMin = 0;
			endTemp = 152;
			startTemp = 152;
			minutes = 60;
			method = "infusion";
			type = "alpha";

		}
	
		// getter/setter methods	
		
		public Quantity getDecoctVol() {
			return decoctVol;
		}

		public void setDecoctVol(Quantity decoctVol) {
			this.decoctVol = decoctVol;
		}
	
		public String getDirections() {
			return directions;
		}

		public void setDirections(String directions) {
			this.directions = directions;
		}

		public double getEndTemp() {
			return endTemp;
		}

		public void setEndTemp(double endTemp) {
			this.endTemp = endTemp;
		}

		public Quantity getInfuseVol() {
			return infuseVol;
		}

		public void setInfuseVol(Quantity infuseVol) {
			this.infuseVol = infuseVol;
		}

		public String getMethod() {
			return method;
		}

		public void setMethod(String method) {
			this.method = method;
		}

		public int getMinutes() {
			return minutes;
		}

		public void setMinutes(int minutes) {
			this.minutes = minutes;
		}

		public int getRampMin() {
			return rampMin;
		}

		public void setRampMin(int rampMin) {
			this.rampMin = rampMin;
		}

		public double getStartTemp() {
			return startTemp;
		}

		public void setStartTemp(double startTemp) {
			this.startTemp = startTemp;
		}

		public String getType() {
			return type;
		}
		
		public void setType(String s){ type = s; }
	}

	public void addStep(String t, double st, double et, String tu, String m, int min,
			int rmin) {
		MashStep step = new MashStep(t, st, et, tu, m, min, rmin);
		steps.add(step);		
		calcMashSchedule();
	}
	
	public void addStep(){
		MashStep step = new MashStep();
		steps.add(step);
		calcMashSchedule();
	}
	
	public void delStep(int i){
		if (steps.size()>i && !steps.isEmpty()){			
			steps.remove(i);
			calcMashSchedule();
		}
			
	}

	// set methods:
	public void setMaltWeight(double mw) {	maltWeightLbs = mw;	}
	public void setMashRatio(double mr){ 
		mashRatio = mr; 
		calcMashSchedule();
	}	
	
	public void setMashRatioU(String u){ 
		mashRatioU = u;
		calcMashSchedule();
	}
	
	public void setMashVolUnits(String u){ 
		volUnits = u;
		calcMashSchedule();
	}
	
	public void setMashTempUnits(String t){
		tempUnits = t;
		calcMashSchedule();
	}
	
	public void setGrainTemp(double t){
		grainTemp = t;
		calcMashSchedule();
	}

	
	// get methods:
	public String getMashVolUnits(){ return volUnits; }
	public String getMashTempUnits(){ return tempUnits; }
	public int getMashTotalTime(){ return totalTime; }
	public double getGrainTemp() { return grainTemp; }
	public String getMashTotalVol() {
		Quantity q = new Quantity();
		q.setQuantity(null, "qt", volQts);
		double d = q.getValueAs(volUnits);
		String s = df1.format(d).toString() + " " + volUnits;
		return s;
	
	}
	
	
	// mash step methods:
	public int setStepType(int i, String t){
		if (steps.size() < i || steps.isEmpty())
			return -1;
		MashStep ms = (MashStep)steps.get(i);
		ms.setType(t);
		ms.setStartTemp(calcStepTemp(t));
		ms.setEndTemp(calcStepTemp(t));
		return 0;
	}
	
	public String getStepType(int i) {
		if (steps.size() < i || steps.isEmpty())
			return "";
		MashStep ms = (MashStep)steps.get(i);
		return ms.getType();		
	}
	
	
	public String getStepDirections(int i){
		return ((MashStep)steps.get(i)).getDirections();
	}
	
	public void setStepMethod(int i, String m){
		((MashStep)steps.get(i)).setMethod(m);
		calcMashSchedule();
	}
	
	public String getStepMethod(int i) {
		return ((MashStep)steps.get(i)).getMethod();	
	}
	
	public void setStepStartTemp(int i, double t){		
		if (tempUnits.equals("C")){
			t = cToF(t);			
		}				
		((MashStep)steps.get(i)).setStartTemp(t);
		((MashStep)steps.get(i)).setEndTemp(t);
		((MashStep)steps.get(i)).setType(calcStepType(t));
		
		calcMashSchedule();
		
	}
	
	public double getStepStartTemp(int i) {
		if (tempUnits.equals("C"))
			return fToC(((MashStep)steps.get(i)).getStartTemp());
		else
			return ((MashStep)steps.get(i)).getStartTemp();	
	}
	
	public void setStepEndTemp(int i, int t){
		if (tempUnits.equals("C"))
			((MashStep)steps.get(i)).setEndTemp(cToF(t));
		else
			((MashStep)steps.get(i)).setEndTemp(t);
		calcMashSchedule();
	}
	
	public double getStepEndTemp(int i) {
		if (tempUnits.equals("C"))
			return fToC(((MashStep)steps.get(i)).getEndTemp());
		else
			return ((MashStep)steps.get(i)).getEndTemp();	
		
	}
	
	public void setStepRampMin(int i, int m){
		((MashStep)steps.get(i)).setRampMin(m);
	}
	
	public int getStepRampMin(int i) {
		return ((MashStep)steps.get(i)).getRampMin();	
	}
	
	public void setStepMin(int i, int m){
		((MashStep)steps.get(i)).setMinutes(m);
	}
	
	public int getStepMin(int i) {
		return ((MashStep)steps.get(i)).getMinutes();	
	}

	public double getStepInfuseVol(int i) {
		return ((MashStep)steps.get(i)).getInfuseVol().getValue();	
	}
	
	public double getStepDecoctVol(int i) {
		return ((MashStep)steps.get(i)).getDecoctVol().getValue();	
	}
	
	
	
	
	public int getStepSize(){
		return steps.size();
	}
	
	// Introducing: the big huge mash calc method!

	public void calcMashSchedule() {
		// Method to run through the mash table and calculate values

		
		double strikeTemp = 0;
		double targetTemp = 0;
		double endTemp = 0;
		double waterAddedQTS = 0;
		double waterEquiv = 0;
		double mr = mashRatio;
		double currentTemp = grainTemp;

		double displTemp = 0;
		double tunLoss = 0; // figure out a better way to do this, eg: themal mass
		double decoct = 0;
		int totalMashTime = 0;
		double totalSpargeTime = 0;
		double mashWaterQTS = 0;
		double mashVolQTS = 0;
		String stepType;
		int numSparge = 0;
		String mashVolU; // gotta define this somewhere

		// convert mash ratio to qts/lb if in l/kg
		if (mashRatioU.equalsIgnoreCase("l/kg")) {
			mr *= 0.479325;
		}

		// convert CurrentTemp to F
		if (mashTempU == "C") {
			currentTemp = cToF(currentTemp);
			tunLoss = tunLoss / 1.8;
		}

		// perform calcs on first record	  
		if (steps.isEmpty())
			return;
		
		// sort the list
		Collections.sort(steps, new stepComparator());

		MashStep stp = ((MashStep) steps.get(0));
		targetTemp = stp.startTemp;
		endTemp = stp.endTemp;
		strikeTemp = calcStrikeTemp(targetTemp, currentTemp, mr, tunLoss);
		waterAddedQTS = mr * maltWeightLbs;
		waterEquiv = maltWeightLbs * (0.192 + mr);
		mashVolQTS = calcMashVol(maltWeightLbs, mr);
		totalMashTime += stp.minutes;
		mashWaterQTS += waterAddedQTS;
		stp.infuseVol.setQuantity(null, "qt", waterAddedQTS);
		stp.method = "infusion";

		// subtract the water added from the Water Equiv so that they are correct when added in the next part of the loop
		waterEquiv -= waterAddedQTS;

		// Updated the water added

		if (tempUnits == "C")
			strikeTemp = fToC(strikeTemp);
		stp.directions = "Mash in with " + df1.format(stp.infuseVol.getValueAs(volUnits)) + " " + volUnits
				+ " of water at " + df1.format(strikeTemp) + " " + tempUnits;

		// set TargetTemp to the end temp
		targetTemp = stp.endTemp;

		for (int i = 1; i < steps.size(); i++) {
			stp = ((MashStep) steps.get(i));
			currentTemp = targetTemp; // switch
			targetTemp = stp.startTemp;

			// do calcs
			if (stp.method.equals("infusion")) { // calculate an infusion step
				decoct = 0;
				waterEquiv += waterAddedQTS; // add previous addition to get WE
				strikeTemp = 212; // boiling water -- TODO: add from options

				// Updated the water added
				waterAddedQTS = calcWaterAddition(targetTemp, currentTemp,
						waterEquiv, 212);
				stp.infuseVol.setQuantity(null, "qt", waterAddedQTS);
				if (tempUnits == "C")
					strikeTemp = 100;
				stp.directions = "Add " + df1.format(stp.infuseVol.getValueAs(volUnits)) + " " + volUnits
						+ " of water at " + df1.format(strikeTemp) + " " + tempUnits;

				mashWaterQTS += waterAddedQTS;
				mashVolQTS += waterAddedQTS;

			}

			else if (stp.method.equals("decoction")) { // calculate a decoction step

				waterEquiv += waterAddedQTS; // add previous addition to get WE
				waterAddedQTS = 0;
				strikeTemp = 212; // boiling water

				// Calculate volume (qts) of mash to remove
				decoct = calcDecoction(targetTemp, currentTemp, waterEquiv, 212);

				// Updated the decoction, convert to right units & make directions
				double decoctConv = decoct;
				stp.directions = "Remove " + df1.format(decoctConv) + " " + volUnits
						+ " of mash, boil, and return to mash.";

			} else if (stp.method.equals("direct")) { // calculate a direct heat step
				decoct = 0;
				waterEquiv += waterAddedQTS; // add previous addition to get WE
				waterAddedQTS = 0;
				stp.directions = "Add direct heat until mash reaches " + displTemp
						+ " " + tempUnits + ".";
			}

			if (stp.method.equals("sparge")) { // Count it
				numSparge++;
			} else
				totalMashTime += stp.minutes;

			stp.infuseVol.setQuantity(null, "qt", waterAddedQTS);
			stp.decoctVol.setQuantity(null, "qt", decoct);

			// set target temp to end temp for next step
			targetTemp = stp.endTemp;

		} // while not eof

		waterEquiv += waterAddedQTS; // add previous addition to get WE
		totalTime = totalMashTime;
		volQts = mashVolQTS;

		// water use stats:
		absorbedQTS = maltWeightLbs * 0.55; // figure from HBD
		// spargeTotalQTS = (r.getPreBoilVol("qt")) - (mashWaterQTS - absorbedQTS);
		totalWaterQTS = spargeTotalQTS + mashWaterQTS;
		// chillShrink = r.getPostBoilVol("gal") * 0.03;
		
		// TODO: sparge stuff should get figured here:		 

	}
	

	// private methods:
	private String calcStepType(double temp) {
		String stepType = "none";
		// less than 90, none
		// 90 - 103 - acid
		if (temp > 89 && temp < 104)
			stepType = "acid";
		// 103 - 110 - gluten
		else if (temp < 111)
			stepType = "gluten";
		// 110 - 135 protein
		else if (temp < 136)
			stepType = "protein";
		// 135 - 149 beta
		else if (temp < 150)
			stepType = "beta";
		// 150-160 alpha
		else if (temp < 161)
			stepType = "alpha";
		// 160-169, mashout
		else if (temp < 170)
			stepType = "mashout";
		// over 170, sparge
		else if (temp >= 170)
			stepType = "sparge";

		return stepType;
	}

	private double calcStepTemp(String stepType) {
		float stepTempF = 0;
		if (stepType == "acid")
			stepTempF = 95;
		else if (stepType == "gluten")
			stepTempF = 105;
		else if (stepType == "protein")
			stepTempF = 122;
		else if (stepType == "beta")
			stepTempF = 149;
		else if (stepType == "alpha")
			stepTempF = 155;
		else if (stepType == "mashout")
			stepTempF = 168;
		else if (stepType == "sparge")
			stepTempF = 170;

		return stepTempF;
	}

	double calcMashVol(double grainWeightLBS, double ratio) {
		// given lbs and ratio, what is the volume of the grain in quarts?
		// note: this calc is for the first record only, and returns the heat equivalent of
		// grain + water added for first infusion
		// HBD posts indicate 0.32, but reality is closer to 0.42

		return (grainWeightLBS * (0.42 + ratio));
	}

	double calcStrikeTemp(double targetTemp, double currentTemp, double ratio,
			double tunLossF) {
		// calculate strike temp
		// Ratio is in quarts / lb, TunLoss is in F
		return (targetTemp + 0.192 * (targetTemp - currentTemp) / ratio)
				+ tunLossF;
	}

	double calcWaterAddition(double targetTemp, double currentTemp,
			double mashVol, double boilTempF) {
		// calculate amount of boiling water to add to raise mash to new temp
		return (mashVol * (targetTemp - currentTemp) / (boilTempF - targetTemp));
	}

	double calcDecoction(double targetTemp, double currentTemp, double mashVol,
			double boilTempF) {
		// calculate amount of grain/mash to remove and boil to raise mash temp to
		// a target temperature
		// returns percent

		double dectPercent = (targetTemp - boilTempF)
				/ (currentTemp - boilTempF);
		dectPercent = 1 - dectPercent;
		return (mashVol * dectPercent);
	}

	private double fToC(double tempF) {
		// routine to convert basic F temps to C
		return (5 * (tempF - 32)) / 9;
	}

	private double cToF(double tempC) {
		// routine to convert Celcius to Farenheit
		return ((tempC * 9) / 5) + 32;
	}
	
	
	 private class batchSparge {
	    int charges;
	    double charge[];
	    double temp[];
	    double collect[];
	 };

	 void calc_batch_sparge(batchSparge bs, double absorbedQts, double usedQts, double total_collectQts) {
	   int i=0;
	   double collect = total_collectQts / bs.charges;

	   // all units in Qts!!!

	   // is there more in the tun than we need?
	   if (collect < usedQts - absorbedQts) {
	     bs.charge[0] = 0;
	     bs.collect[0] = usedQts - absorbedQts;
	     total_collectQts = total_collectQts - bs.collect[0];
	   }
	   else {
	     bs.charge[0] = collect - (usedQts - absorbedQts);
	     bs.collect[0] = collect;
	     total_collectQts = total_collectQts - bs.collect[0];
	   }

	   // do we need any more steps?
	   if (bs.charges == 1) return;

	   collect = total_collectQts / (bs.charges - 1);
	   for (i=1; i<bs.charges; i++) {
	     bs.charge[i] = collect;
	     bs.collect[i] = collect;
	   }

	   return;
	 };
	 
	 public String toXml() {
	
		StringBuffer sb = new StringBuffer();
		sb.append("  <MASH>\n");
		for (int i = 0; i < steps.size(); i++) {
			MashStep st = (MashStep) steps.get(i);
			sb.append("    <ITEM>\n");
			sb.append("      <TYPE>" + st.type + "</TYPE>\n");
			sb.append("      <TEMP>" + st.startTemp + "</TEMP>\n");
			sb.append("      <END_TEMP>" + st.endTemp + "</END_TEMP>\n");
			sb.append("      <METHOD>" + st.method + "</METHOD>\n");
			sb.append("      <DIRECTIONS>" + st.directions + "</DIRECTIONS>\n");
			sb.append("    </ITEM>\n");
		}

		sb.append("  </MASH>\n");
		return sb.toString();
	}
	
	 /**
		 * 
		 * @author aavis
		 *
		 * step comparator to help sort mash steps
		 * 
		 */
		public class stepComparator implements Comparator {			
			public int compare(Object a, Object b) {
				Double a1 = new Double(((MashStep)a).getStartTemp());
				Double b1 = new Double(((MashStep)b).getStartTemp());				
				int result = a1.compareTo(b1);
				return (result == 0 ? -1 : result);
				
			}
		}

}